test/ability_leaf_guard.c

```diff
@@ -0,0 +1,48 @@
+#include "global.h"
+#include "test_battle.h"
+
+SINGLE_BATTLE_TEST("Leaf Guard prevents non-volatile status conditions in sun")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_WILL_O_WISP].effect == EFFECT_WILL_O_WISP);
+        PLAYER(SPECIES_LEAFEON) { Ability(ABILITY_LEAF_GUARD); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_SUNNY_DAY); MOVE(opponent, MOVE_WILL_O_WISP); }
+    } SCENE {
+        ABILITY_POPUP(player, ABILITY_LEAF_GUARD);
+        NOT ANIMATION(ANIM_TYPE_MOVE, MOVE_WILL_O_WISP, opponent);
+        MESSAGE("It doesn't affect Leafeon…");
+        NOT STATUS_ICON(player, burn: TRUE);
+    }
+}
+
+SINGLE_BATTLE_TEST("Leaf Guard prevents non-volatile status conditions from items")
+{
+    GIVEN {
+        ASSUME(gItems[ITEM_FLAME_ORB].holdEffect == HOLD_EFFECT_FLAME_ORB);
+        PLAYER(SPECIES_LEAFEON) {Ability(ABILITY_LEAF_GUARD); Item(ITEM_FLAME_ORB); };
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_SUNNY_DAY); }
+    } SCENE {
+        NOT MESSAGE("Leafeon was burned!");
+        NOT STATUS_ICON(player, burn: TRUE);
+    }
+}
```

Comment:
I would add tests for each status condition tbh, to make sure it indeed prevents *all* non-volatile status conditions.

Reply:
You could `PARAMETRIZE` each move and then add a switch to change the message between messages